### DISCUSSION TITLE
Problems running Buildings library on ARM chip

### DISCUSSION THREAD

**Discussion by vtnate** [OpenModelica/OpenModelica#11751] [Q&A] [answered]
I'm trying to run `Buildings.Experimental.DHC.Examples.Combined.SeriesConstantFlow` and getting

> ... loading "tab1" from "/usr/local/Modelica/Library/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissOffice_20190916.mos"
... loading "tab1" from "/usr/local/Modelica/Library/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissResidential_20190916.mos"
... loading "tab1" from "/usr/local/Modelica/Library/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissHospital_20190916.mos"
SeriesConstantFlow_functions.c:21114: Invalid root: (0)^(-1.5)
simulation terminated by an assertion at initialization
Simulation process failed. Exited with code 255.

I'm using an M1 Mac laptop, running a VM with Ubuntu 22.04.3, OM 1.22.1, and MBL 10. This model works fine for my coworker with an Intel Mac (and in the Buildings team CI testing, also on Intel), so I'm becoming suspicious that the ARM chip is not playing nice with this model. Is there something in OM that the ARM chip uses differently?

Thanks for all your help,

Nathan

**Comment by casella** [comment]
@sjoelund would you like to comment on that?

**Reply by sjoelund** [reply]
There's not really any code that is different on ARM than on Intel. But similar errors may occur also when running on 2 different Intel chips due to using different instructions for floating point operations. I am guessing that the 0 here is `-0` which you cannot take the root of.

**Reply by casella** [reply]
As far as I understand, you cannot raise zero to any negative power, because that gives you infinity 😃

**Comment by vtnate** [comment]
So is there a way for OM to handle the different floating point operations? Surely just letting it not work for some people isn't the answer, is it? How can we get ARM chips to run the same models as Intel chips, if indeed that is the issue here?

**Reply by sjoelund** [reply]
They are running LAPACK versions/etc that compute slightly different results.  You will not get the identical results even on different versions of Ubuntu on the same machine. It's usually not a problem, but sometimes you can hit problems like this where you get exactly zero on one system and a very small number on another.

**Reply by vtnate** [reply]
Thanks for considering this. Is there anything we can do so this model will work for ARM computers?

**Reply by vtnate** [reply]
Or if it's not an ARM problem, how can we get this model to simulate everywhere the same way? Is there a setting in OM that I don't know about that would impact this?

**Comment by dershow** [comment]
It's not really an ARM problem if you are running in a virtual machine that is emulating an Intel processor.  
I'm not familiar with this library, but I did build a native OM for Mac ARM (OM 1.23.0-dev-214) so decided to try it out.  
I also get an error, although it's different from what you see:

> ... loading "tab1" from "/Users/adershowitz/.openmodelica/libraries/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissOffice_20190916.mos" 
... loading "tab1" from "/Users/adershowitz/.openmodelica/libraries/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissResidential_20190916.mos" 
... loading "tab1" from "/Users/adershowitz/.openmodelica/libraries/Buildings 10.0.0/Resources/Data/Experimental/DHC/Loads/Examples/SwissHospital_20190916.mos" 
Model error: Argument of log(dT2 / dT1) was -1 should be > 0
Non-Linear Solver try to handle a problem with a called assert.
Model error: Argument of log(dT2 / dT1) was -1 should be > 0
Non-Linear Solver try to handle a problem with a called assert.
Model error: Argument of log(dT2 / dT1) was -1 should be > 0
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
Division by zero roughness / diameter in function context
Non-Linear Solver try to handle a problem with a called assert.
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
Division by zero roughness / diameter in function context
While solving non-linear system an assertion failed during initialization.
The non-linear solver tries to solve the problem that could take some time.
It could help to provide better start-values for the iteration variables.
For more information simulate with -lv LOG_NLS_V
... loading "TStep" from "tmp/temperatureResponseMatrix/cb5a516a14d0ff25ffce0ba805d0acd30c68bbadTStep.mat" 
Variable "TStep" not found on file "tmp/temperatureResponseMatrix/cb5a516a14d0ff25ffce0ba805d0acd30c68bbadTStep.mat". 
Failed to solve the initialization problem with global homotopy with equidistant step size.
Unable to solve initialization problem.
simulation terminated by an assertion at initialization

I'm not sure if this is at all helpful, since I don't know what this particular library should do with default values, version 1.23.0 vs 1.22.1
This looks to me like perhaps different initial values might help?

**Reply by casella** [reply]
You could try to run with -lv=LOG_NLS_V and see what are the initial guess values for the Newton solvers. It seems that some of those values are really not good.

**Reply by vtnate** [reply]
Thanks @casella, that's helpful to have instructions for troubleshooting. I ran it using OMEdit, so I just checked the box for LOG_NLS_V and got the same error with this additional line: `free non-linear system solvers`. It also told me there were `80 non-linear systems`. Unfortunately I don't know what any of that means 😞 🙂

**Reply by vtnate** [reply]
There was no other output. No initial guess values or anything. Exact same error message as my initial post, with the addition of those 2 lines mentioned above.

**Reply by dershow** [reply]
I see from your comments that you are actually running ARM Ubuntu so are not emulating.  So, this might be ARM related.

I a run, with LOG_NLS_V on the Mac build, and I'm not really sure what part of the output might help address the question, but I've attached it below:
[SeriesConstantFlow.log](https://github.com/OpenModelica/OpenModelica/files/13876780/SeriesConstantFlow.log)

**Comment by vtnate** [comment]
Perhaps this is more indirectly related to ARM. I've been troubleshooting with the Buildings team from LBNL and we've narrowed it down to an issue that only exists on Ubuntu 22. Ubuntu 20 appears to run this test as expected. Unfortunately, only Ubuntu 22 is compatible with ARM chips, such as any recent Apple computer.

I'm at a loss as how to move forward at this point.

**Reply by sjoelund** [reply]
We have been running OpenModelica on 32-bit and 64-bit ARM chips for quite a few years now. Or is it only Ubuntu 22 that runs on Apple Silicon that is the problem?

Otherwise, you could perhaps run docker on Ubuntu 22.04, running a docker container based on Ubuntu 20.04... But I would guess you could get similar issues in the future from just changing initial guess values.

**Reply by vtnate** [reply]
Thanks for continuing to think about this, Martin. The problem appears to be that Ubuntu only started supporting ARM chips with version 22, and this model only fails on Ubuntu 22. Perhaps the upcoming Ubuntu 24 will magically resolve my problem, but I'm not holding my breath for that! 😆

**Reply by sjoelund** [reply]
Pretty sure we ran Ubuntu 16.04 and 18.04 on arm at some point. Maybe even earlier...
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal
$ uname -m
aarch64
```

**Reply by vtnate** [reply]
Fascinating. Can you point me towards an Ubuntu 20 ISO I can download for ARM? Maybe my googling skills are weak today.

**Reply by sjoelund** [reply]
I usually install my ARM systems through https://www.armbian.com/ since so many ARM systems have hardware that requires some extra firmware or packages or bootloaders to work properly.

But Ubuntu has some:
https://cdimage.ubuntu.com/releases/focal/release/ 20.04
https://cdimage.ubuntu.com/releases/bionic/release/ 18.04

**Reply by vtnate** [reply]
Marvelous, thank you! I've searched all over the Ubuntu website and Stack Overflow and never found those links. I'll try with 20.04 and maybe this will work for me.

**Comment by vtnate** [comment, marked as answer]
I was able to successfully run the model with Ubuntu 20. It's not at all clear to me what changed in Ubuntu 22 to make this model fail, but at least I have a workaround now.

This issue is closed, as far as I'm concerned. Thank you for your assistance @sjoelund
